parser/yaml/yaml.go

```diff
@@ -0,0 +1,831 @@
+//  Copyright 2022 Elphas Tori
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+//Package yaml contains the implementation of a YAML parser.
+package yaml
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+
+	"github.com/katydid/katydid/parser"
+)
+
+//YamlParser is a parser for YAML
+type YamlParser interface {
+	parser.Interface
+	//Init initialises the parser with a byte buffer containing YAML.
+	Init(buf []byte) error
+	Reset() error
+}
+
+//NewYamlParser returns a new YAML parser.
+func NewYamlParser() YamlParser {
+	return &yamlParser{
+		state: state{
+			firstObjectValue: true,
+		},
+		stack: make([]state, 0, 10),
+	}
+}
+
+func (s *yamlParser) Init(buf []byte) error {
+	s.state = state{
+		firstObjectValue: true,
+		buf:              buf,
+	}
+	// empty stack in case this is an initialization due to a reset
+	s.stack = s.stack[:0]
+	if err := s.skipEmptyLines(); err != nil {
+		return err
+	}
+	line, err := s.peekNextLine()
+	if err != nil {
+		return err
+	}
+
+	if line.isArray {
+		if err := s.scanValue(); err != nil {
+			return err
+		}
+		s.inArray = true
+		s.firstArrayValue = true
+		s.buf = s.buf[s.startValueOffset:s.endValueOffset]
+		s.offset = 0
+	}
+	return nil
+}
+
+func (s *yamlParser) scanValue() error {
+	line, err := s.peekNextLine()
+	if err != nil {
+		return err
+	}
+	if line.isArray {
+		return s.scanArray()
+	}
+
+	if line.hasValue { // tighten this condition
+		value := []byte(strings.TrimSpace(string(line.value)))
+		if len(value) > 0 {
+			return s.scanString()
+		}
+	}
+
+	return s.scanString()
+	// return s.expected("objects not supported")
```

Comment:
maybe we can remove the comment? and a few others below, or do you want to maybe expand on them later, if so let's add that to the comment.

Reply:
Removed redundant comments in YAML parser implementation